### PR TITLE
docs(recipes): pwa and modules

### PR DESCRIPTION
**Pull request by Francois-Esquire**
Documentation recipe for using PWA with modules.

## Description

This is a recipe that goes over PWA enhancements we can add to our component.

## Motivation and Context

The recipe adds a guide to setup PWA in One App and use some of the facilities provided.

## How Has This Been Tested?

N/A

## Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation (adding or updating documentation)
- [ ] Dependency update

## Checklist:

- [ ] My change requires a change to the documentation and I have updated the documentation accordingly.
- [ ] These changes should be applied to a maintenance branch.
- [ ] This change requires cross browser checks.
- [ ] Performance tests should be ran against the server prior to merging.
- [ ] This change impacts caching for client browsers.
- [ ] This change impacts HTTP headers.
- [ ] This change adds additional environment variable requirements for One App users.
- [ ] I have added the Apache 2.0 license header to any new files created.

## What is the Impact to Developers Using One App?

Documentation

### CODE REVIEW

File: docs/recipes/PWA.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+* [Offline Support](#offline-support)
+* [App Installation](#app-installation)
+
+## Overview
+
+In this compilation of mini recipes well be getting our feet wet with Progressive Web Apps
+(PWA) in One App. We'll learn a bit about the service worker and how to use the web
+manifest with One App. Each mini recipe below will cover UX issues and the technical
+solutions to refine your One App PWA.
+If any of these terms are new, we have links ready to get you familiar below in this section.
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:23]: Nitpick on why the line break here.

File: docs/recipes/PWA.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+- [Web App Manifest](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps)
+- [App Installs](https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Developer_guide/Installing)
+- [Online/Offline](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/Online_and_offline_events)
+- [Permissions](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API)
+- [Cache](https://developer.mozilla.org/en-US/docs/Web/API/Cache)
+- [Cache Storage](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage)
+
+**Examples**
+
+* [`frank-lloyd-root` v0.0.3](frank-lloyd-root) sample module configures and enables PWA.
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:43]: Maybe `frank-lloyd-root@0.0.3` it looks weird (visually) in the readme currently with `frank-lloyd-root` and then v0.0.3 following it and that will translate to the docs site as well.

File: docs/recipes/PWA.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+export ONE_SERVICE_WORKER=1 || true
+```
+
+This is set to `false` by default.
+
+**Root Module**
+
+Next step is to add it to your `appConfig` on your root module. The `pwa` key has
+a few options that we can use to turn on the service worker and create a web manifest.
+Here is a sample configuration:
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:63]:
```suggestion
Next step is to add it to your `appConfig` in your root module. The `pwa` key has
a few options that we can use to turn on the service worker and create a web manifest.
Here is a sample configuration:
```

File: docs/recipes/PWA.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
+<!--ONE-DOCS-HIDE start-->
+[👈 Return to Overview](./README.md)
+<!--ONE-DOCS-HIDE end-->
+
+[one-app-config]: ../api/modules/App-Configuration.md
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:5]:
```suggestion
[one-app-config]: ../api/modules/App-Configuration.md#pwa
```

File: docs/recipes/PWA.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+load in a PWA ready One App. The service worker is installed automatically
+and the web manifest we've configured will be added to the rendered HTML.
+Under the hood, [`@americanexpress/one-service-worker`][one-service-worker]
+drives the service worker that One App is using.
+What you will get out of the box is caching for the `one-app` client, your modules,
+including language packs bundled in your package; One App also supports
+offline navigation, which means that when a user requests one of our pages when offline,
+`one-app` client and modules will both load up and it is because what we have already
+cached is given back to the browser when the network is down. Try testing what your modules
+do when offline.
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:111]: This is starting to sound like a run-on sentence

File: docs/recipes/PWA.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+
+**Your development browser matters**
```
Comment (JamesSingleton) [inline docs/recipes/PWA.md:113]: Do we have any recommendations? Do we want to recommend anything?

File: docs/recipes/PWA.md [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+
+In this compilation of mini recipes well be getting our feet wet with Progressive Web Apps
+(PWA) in One App. We'll learn a bit about the service worker and how to use the web
+manifest with One App. Each mini recipe below will cover UX issues and the technical
+solutions to refine your One App PWA.
+If any of these terms are new, we have links ready to get you familiar below in this section.
+We recommend going through and reviewing the links before using PWA if you haven't
+already done so.
+
+**Links**
```
Comment (JAdshead) [inline docs/recipes/PWA.md:27]: what about making this "prerequisites"

File: docs/recipes/PWA.md [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+## Setup
+
+Let's start by configuring One App to enable the service worker.
+
+**One App**
+
+First thing is we need to set `ONE_SERVICE_WORKER` before we run One App:
+
+```bash
+export ONE_SERVICE_WORKER=1 || true
```
Comment (JAdshead) [inline docs/recipes/PWA.md:54]: keep it simple, 
```suggestion
export ONE_SERVICE_WORKER=true
```

File: docs/recipes/PWA.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+            sizes: '96x96',
+            type: 'image/png',
+          },
+        ],
+      },
+    },
+  };
+}
+```
+
```
Comment (JAdshead) [inline docs/recipes/PWA.md:97]: can you add a description of what this config will achieve

File: docs/recipes/PWA.md (hunk 5 above)
Comment (JAdshead) [reply docs/recipes/PWA.md:111]: maybe make list benefits as bullet points

File: docs/recipes/PWA.md (hunk 6 above)
Comment (JAdshead) [reply docs/recipes/PWA.md:113]: can you expand on this

File: docs/recipes/PWA.md [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+the service worker, the web manifest and the cache.
+Skip to the next paragraph when you've selected a browser.
+
+**Start**
+
+When you're ready to start, here are a few mini recipes to outfit your
+Holocron modules with PWA enhancements.
+
+## Offline Support
+
```
Comment (JAdshead) [inline docs/recipes/PWA.md:125]: what about listing the positives and negatives

File: docs/recipes/PWA.md [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+      // inform user
+      // collect form submission
+      // collect analytics/errors/reports
+    }
+  }, [offline]);
+  return <p>Offline Mode</p>;
+}
+
+export default function MainModule() {
+  if (useOffline()) {
```
Comment (JAdshead) [inline docs/recipes/PWA.md:193]: not sure hooks should be used as conditional 
```suggestion
 const isOffline = useOffline();
  if (isOffline) {
```

File: docs/recipes/PWA.md [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+Once we know if/when we're on or off the network, we can make our modules
+ready for offline and tweak its behavior appropriately.
+
+```jsx
+import React from 'react';
+import { isOffline } from '@americanexpress/one-service-worker';
+
+function useOffline() {
+  const [offline, setOffline] = React.useState(false);
+  React.useLayoutEffect(() => {
```
Comment (JAdshead) [inline docs/recipes/PWA.md:153]: might also be worth justifying why they would need to use `useLayoutEffect`  over `useEffect`

**Review by JAdshead — CHANGES_REQUESTED** [review]

File: docs/recipes/PWA.md (hunk 1 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:23]: No particular reason, I'll update.

File: docs/recipes/PWA.md (hunk 5 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:111]: Good points made, I'll update.

File: docs/recipes/PWA.md (hunk 6 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:113]: I did want to talk about Chrome and using the browsers dev tools - I can bring that in the article.

File: docs/recipes/PWA.md (hunk 12 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:210]: I added a comments in the offline example (b912ca5)

File: docs/recipes/PWA.md (hunk 10 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:164]: Added a list in b912ca5

File: docs/recipes/PWA.md (hunk 9 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:116]: I added comments to the config items and expanded the config (b912ca5)

File: docs/recipes/PWA.md (hunk 7 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md]: Updated (b912ca5)

File: docs/recipes/PWA.md (hunk 6 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md:153]: I included Chrome in the paragraph (b912ca5)

File: docs/recipes/PWA.md (hunk 5 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md]: rewrote that paragraph (b912ca5), let me know if it got worse lol

File: docs/recipes/PWA.md (hunk 2 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md]: updated (b912ca5)

File: docs/recipes/PWA.md (hunk 1 above)
Comment (Francois-Esquire) [reply docs/recipes/PWA.md]: Cleaned this up in b912ca5

**Review by JAdshead — APPROVED** [review]

**Review by JamesSingleton — APPROVED** [review]

**Pull request merged**